### PR TITLE
[ie/neteasemusic] Add additional quality formats

### LINKED ISSUE #9196: [NeteaseMusic] Extract flac audio
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a site-specific feature
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

China

### Example URLs

https://music.163.com/#/song?id=1952853555

### Provide a description that is worded well enough to be understood

Netease music extractor is unable to extract high quality audio from API.

The reason (AFAIK) is that the extractor [searches for specific media IDs](https://github.com/yt-dlp/yt-dlp/blob/882e3b753c79c7799ce135c3a5edb72494b576af/yt_dlp/extractor/neteasemusic.py#L25), which does not cover all possible IDs.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['--list-formats', '--cookies', '/home/hafeoz/.ytdlp-cookies.txt', '-vU', 'https://music.163.com/#/song?id=1952853555']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.12.30 from yt-dlp/yt-dlp [f10589e34]
[debug] Python 3.11.6 (CPython x86_64 64bit) - Linux-6.7-arch3-1-x86_64-with-glibc2.38 (OpenSSL 3.2.0 23 Nov 2023, glibc 2.38)
[debug] exe versions: ffmpeg 6.1.1 (setts), ffprobe 6.1.1, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.12.0, certifi-2024.02.02, mutagen-1.47.0, requests-2.31.0, secretstorage-3.3.3, sqlite3-3.45.1, urllib3-1.26.18
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests
[debug] Loaded 1798 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: stable@2023.12.30 from yt-dlp/yt-dlp
yt-dlp is up to date (stable@2023.12.30 from yt-dlp/yt-dlp)
[netease:song] Extracting URL: https://music.163.com/#/song?id=1952853555
[netease:song] 1952853555: Downloading song info
[netease:song] 1952853555: Downloading song URL info: bitrate 128002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 192002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 320002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading lyrics data
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 1952853555:
ID     EXT RESOLUTION │ FILESIZE PROTO │ VCODEC  ACODEC   ABR ASR
─────────────────────────────────────────────────────────────────
bMusic mp3 unknown    │  2.93MiB http  │ unknown unknown 128k 48k
mMusic mp3 unknown    │  4.40MiB http  │ unknown unknown 192k 48k
hMusic mp3 unknown    │  7.33MiB http  │ unknown unknown 320k 48k
```

### PR DESCRIPTION
**Pull request by hafeoz**
### Description of your *pull request* and other information

This PR adds several format IDs to the extractor, allowing high quality audio to be extracted from Netease Music.

The list of format IDs are constructed based on API response of https://music.163.com/#/song?id=1952853555, and are cross-checked with [a repo with similar mission based a rough GitHub search](https://github.com/FengLiuFeseliud/pycloudmusic/blob/43078108be2dd7fc635bc2951d31bbed538c251f/pycloudmusic/object/music163.py#L352)

I can't find any tests related to this part of the code, so I didn't change any.

Fixes #9196

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
    I've only skimmed though it - this *seems like* a simple code change so there's probably not much I can break regarding coding conventions.
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by hafeoz** [comment]
Following is the formats available without this PR:
```
[debug] Command-line config: ['--list-formats', '--cookies', '/home/hafeoz/.ytdlp-cookies.txt', '-vU', 'https://music.163.com/#/song?id=1952853555']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.12.30 from yt-dlp/yt-dlp [f10589e34]
[debug] Python 3.11.6 (CPython x86_64 64bit) - Linux-6.7-arch3-1-x86_64-with-glibc2.38 (OpenSSL 3.2.0 23 Nov 2023, glibc 2.38)
[debug] exe versions: ffmpeg 6.1.1 (setts), ffprobe 6.1.1, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.12.0, certifi-2024.02.02, mutagen-1.47.0, requests-2.31.0, secretstorage-3.3.3, sqlite3-3.45.1, urllib3-1.26.18
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests
[debug] Loaded 1798 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: stable@2023.12.30 from yt-dlp/yt-dlp
yt-dlp is up to date (stable@2023.12.30 from yt-dlp/yt-dlp)
[netease:song] Extracting URL: https://music.163.com/#/song?id=1952853555
[netease:song] 1952853555: Downloading song info
[netease:song] 1952853555: Downloading song URL info: bitrate 128002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 192002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 320002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading lyrics data
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 1952853555:
ID     EXT RESOLUTION │ FILESIZE PROTO │ VCODEC  ACODEC   ABR ASR
─────────────────────────────────────────────────────────────────
bMusic mp3 unknown    │  2.93MiB http  │ unknown unknown 128k 48k
mMusic mp3 unknown    │  4.40MiB http  │ unknown unknown 192k 48k
hMusic mp3 unknown    │  7.33MiB http  │ unknown unknown 320k 48k
```

And following is the formats with this patch (same song, logged in SVIP):
```
[netease:song] Extracting URL: https://music.163.com/#/song?id=1952853555
[netease:song] 1952853555: Downloading song info
[netease:song] 1952853555: Downloading song URL info: bitrate 855137
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 1622742
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 128002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 128002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 192002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 320002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading lyrics data
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 1952853555:
ID      EXT  RESOLUTION │  FILESIZE PROTO │ VCODEC  ACODEC   ABR ASR
────────────────────────────────────────────────────────────────────
bMusic  mp3  unknown    │   2.93MiB http  │ unknown unknown 128k 48k
lMusic  mp3  unknown    │   2.93MiB http  │ unknown unknown 128k 48k
mMusic  mp3  unknown    │   4.40MiB http  │ unknown unknown 192k 48k
hMusic  mp3  unknown    │   7.33MiB http  │ unknown unknown 320k 48k
hrMusic flac unknown    │  19.57MiB http  │ unknown unknown 855k 48k
sqMusic flac unknown    │  19.57MiB http  │ unknown unknown 855k 48k
```

And same song without SVIP:
```
[netease:song] Extracting URL: https://music.163.com/#/song?id=1952853555
[netease:song] 1952853555: Downloading song info
[netease:song] 1952853555: Downloading song URL info: bitrate 855137
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 1622742
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 128002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 128002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 192002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading song URL info: bitrate 320002
[netease:song] 1952853555: Checking song URL
[netease:song] 1952853555: Downloading lyrics data
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 1952853555:
ID      EXT RESOLUTION │ FILESIZE PROTO │ VCODEC  ACODEC   ABR ASR
──────────────────────────────────────────────────────────────────
bMusic  mp3 unknown    │  2.93MiB http  │ unknown unknown 128k 48k
lMusic  mp3 unknown    │  2.93MiB http  │ unknown unknown 128k 48k
mMusic  mp3 unknown    │  4.40MiB http  │ unknown unknown 192k 48k
hMusic  mp3 unknown    │  7.33MiB http  │ unknown unknown 320k 48k
hrMusic mp3 unknown    │  7.33MiB http  │ unknown unknown 320k 48k
sqMusic mp3 unknown    │  7.33MiB http  │ unknown unknown 320k 48k
```

And finally `ffprobe` output of logged-in `sqMusic` download sample:
```
Input #0, flac, from '采菊 [1952853555].flac':
  Duration: 00:03:12.01, start: 0.000000, bitrate: 855 kb/s
  Stream #0:0: Audio: flac, 48000 Hz, stereo, s16
```

**Review by pinich — APPROVED** [review]

**Review by pzhlkj6612 — APPROVED** [review]
The new code doesn't break the extraction for the main song in a program with the `--no-playlist` option:

```
[debug] Command-line config: ['--verbose', '--list-formats', '--no-playlist', ...
...
[netease:program] Extracting URL: https://music.163.com/#/program?id=10109051 
[netease:program] 10109051: Downloading program info
[netease:program] Downloading just the video 32591923 because of --no-playlist 
[netease:program] 32591923: Downloading song URL info: bitrate 128000 
[netease:program] 32591923: Checking song URL 
[netease:program] 32591923: Downloading song URL info: bitrate 128000 
[netease:program] 32591923: Checking song URL 
[netease:program] 32591923: Downloading song URL info: bitrate 192000 
[netease:program] 32591923: Checking song URL 
[netease:program] 32591923: Downloading song URL info: bitrate 320000 
[netease:program] 32591923: Checking song URL 
[info] Available formats for 32591923:
ID     EXT RESOLUTION │ FILESIZE PROTO │ VCODEC  ACODEC   ABR ASR
─────────────────────────────────────────────────────────────────
bMusic mp3 unknown    │  2.92MiB http  │ unknown unknown 128k 44k
lMusic mp3 unknown    │  2.92MiB http  │ unknown unknown 128k 44k
mMusic mp3 unknown    │  4.39MiB http  │ unknown unknown 192k 44k
hMusic mp3 unknown    │  7.31MiB http  │ unknown unknown 320k 44k
[info] 32591923: Downloading 4 format(s): hMusic, mMusic, lMusic, bMusic

```

.

**Comment by pzhlkj6612** [comment]
Hi, @hafeoz. Maybe you could apply a small fix for the hint about `--no-playlist` :

```diff
diff --git a/yt_dlp/extractor/neteasemusic.py b/yt_dlp/extractor/neteasemusic.py
index e4876af5c..1877034e5 100644
--- a/yt_dlp/extractor/neteasemusic.py
+++ b/yt_dlp/extractor/neteasemusic.py
@@ -561,7 +561,8 @@ def _real_extract(self, url):
             'timestamp': ('createTime', {self.kilo_or_none}),
         })

-        if not self._yes_playlist(info['songs'] and program_id, info['mainSong']['id']):
+        if not self._yes_playlist(
+            info['songs'] and program_id, info['mainSong']['id'], playlist_label='program', video_label='song'):
             formats = self.extract_formats(info['mainSong'])

             return {

```

Comparison:

```diff
  [netease:program] Extracting URL: https://music.163.com/#/program?id=10109051 
  [netease:program] 10109051: Downloading program info
- [netease:program] Downloading playlist 10109051 - add --no-playlist to download just the video 32591923
+ [netease:program] Downloading program 10109051 - add --no-playlist to download just the song 32591923

```

.

**Comment by bashonly** [comment]
@pzhlkj6612 do you want to open a separate PR for that?

**Comment by pzhlkj6612** [comment]
> @pzhlkj6612 do you want to open a separate PR for that?

Done, #9488.

File: yt_dlp/extractor/neteasemusic.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 
 class NetEaseMusicBaseIE(InfoExtractor):
-    _FORMATS = ['bMusic', 'mMusic', 'hMusic']
+    _FORMATS = ['sqMusic', 'hrMusic', 'lMusic', 'bMusic', 'mMusic', 'hMusic']
```
[the remark this answers is not in the document]
Comment (pzhlkj6612) [reply yt_dlp/extractor/neteasemusic.py:25]: Hi!

That is indeed not a complete list. Do you happen to know the connection between the `xMusic` types and the names in Netease App (or web client)?

[the remark this answers is not in the document]
Comment (timi137137) [reply yt_dlp/extractor/neteasemusic.py:25]: I do know something, but my English is not good enough to explain it clearly.
**You can look [here](https://github.com/yt-dlp/yt-dlp/issues/9196#issuecomment-2185285724).**

The general meaning is that this interface he upgraded, now the call method is different, may need to poll the request to accurately obtain.

The new invocation looks something like this
image

**Comment by hafeoz** [comment]
So it seems like this API is no longer useful (https://github.com/yt-dlp/yt-dlp/issues/9196#issuecomment-2187265378 https://github.com/yt-dlp/yt-dlp/pull/9197#discussion_r1651389618). Closing :(

**Comment by bashonly** [comment]
Feel free to repurpose this PR if you want to implement extraction via the new API

**Comment by hafeoz** [comment]
Sorry I've re-forked (to make PR commit history clean) and GitHub won't let me re-open this PR with the new repo. I've submitted the change as #10258 :)

**Pull request closed without merging**